File: src/data_sources/naver_finance.py
"""
naver_finance.py — Naver Finance data for Korean/Asian companies.

Calls naver_finance.js (Puppeteer subprocess) to scrape finance.naver.com.
Returns structured company financial data in Korean Won (KRW).

Best used when:
  - Target is a Korean KOSPI/KOSDAQ listed company
  - Need Korean market context for a transaction (Widus-style deals)
  - Comparable Korean listed companies for an M&A transaction

USD conversion uses a hardcoded approximate rate; real-time FX is not fetched
(add yfinance USDKRW=X call if precision matters).
"""

import asyncio
import json
import subprocess
from dataclasses import dataclass, field
from pathlib import Path
from typing import Optional

_SCRAPER_JS = Path(__file__).parent / "naver_finance.js"
_PROJECT_ROOT = Path(__file__).parent.parent.parent

# Approximate USD/KRW rate — update periodically or fetch dynamically
_USD_KRW_APPROX = 1_350.0


# ---------------------------------------------------------------------------
# Data structure
# ---------------------------------------------------------------------------

@dataclass
class NaverFinanceResult:
    company_name: str
    stock_code: str = ""
    exchange: str = "KOSPI/KOSDAQ"
    current_price_krw: Optional[int] = None
    page_url: str = ""
    metrics: dict = field(default_factory=dict)  # raw metrics dict from JS
    error: Optional[str] = None

    def current_price_usd(self) -> Optional[float]:
        if self.current_price_krw is None:
            return None
        return self.current_price_krw / _USD_KRW_APPROX

    def to_markdown(self) -> str:
        if self.error:
            return f"_Naver Finance: {self.error}_"

        lines = [
            f"**Naver Finance — {self.company_name}** "
            f"({'KOSPI/KOSDAQ'} · {self.stock_code})",
            "",
        ]

        if self.current_price_krw:
            usd = self.current_price_usd()
            usd_str = f" (~${usd:,.2f} USD)" if usd else ""
            lines.append(f"**Current Price:** ₩{self.current_price_krw:,}{usd_str}")

        # Key metrics
        metric_labels = {
            "per":                    "PER (P/E)",
            "pbr":                    "PBR (P/B)",
            "roe_pct":                "ROE",
            "market_cap_display":     "Market Cap",
            "revenue_display":        "Revenue (TTM)",
            "operating_income_display": "Operating Income",
            "net_income_display":     "Net Income",
            "debt_ratio_pct":         "Debt Ratio",
            "dividend_yield_pct":     "Dividend Yield",
            "eps_krw":                "EPS (KRW)",
        }

        for key, label in metric_labels.items():
            val = self.metrics.get(key)
            if val:
                lines.append(f"**{label}:** {val}")

        if self.page_url:
            lines.append(f"\n_Source: {self.page_url}_")

        return "\n".join(lines)


# ---------------------------------------------------------------------------
# Subprocess call
# ---------------------------------------------------------------------------

async def fetch_naver_finance(company_name: str) -> NaverFinanceResult:
    """
    Fetch company financial data from Naver Finance via Puppeteer.

    Args:
        company_name: Company name in Korean or English (Korean works better
                      for KOSPI/KOSDAQ lookups)

    Returns:
        NaverFinanceResult with stock metrics and KRW pricing
    """
    result = NaverFinanceResult(company_name=company_name)

    if not _SCRAPER_JS.exists():
        result.error = f"Scraper not found: {_SCRAPER_JS}"
        return result

    node_modules = _PROJECT_ROOT / "node_modules" / "puppeteer"
    if not node_modules.exists():
        result.error = "puppeteer not installed — run: npm install (in pe-diligence-tool/)"
        return result

    loop = asyncio.get_event_loop()
    try:
        proc = await loop.run_in_executor(
            None,
            lambda: subprocess.run(
                ["node", str(_SCRAPER_JS), company_name],
                capture_output=True,
                text=True,
                timeout=60,
                cwd=str(_PROJECT_ROOT),
            ),
        )
    except subprocess.TimeoutExpired:
        result.error = "Naver Finance scrape timed out (60s)"
        return result
    except Exception as exc:
        result.error = f"Subprocess error: {exc}"
        return result

    if proc.returncode != 0:
        result.error = (
            proc.stderr.strip()[:200] if proc.stderr.strip()
            else f"Scraper exited {proc.returncode}"
        )
        return result

    try:
        data = json.loads(proc.stdout)
    except json.JSONDecodeError as exc:
        result.error = f"Invalid JSON from scraper: {exc}"
        return result

    result.stock_code         = data.get("stock_code", "")
    result.company_name       = data.get("company_name", company_name)
    result.exchange           = data.get("exchange", "KOSPI/KOSDAQ")
    result.current_price_krw  = data.get("current_price_krw")
    result.page_url           = data.get("page_url", "")
    result.metrics            = data.get("metrics", {})

    return result


# ---------------------------------------------------------------------------
# CLI test
# ---------------------------------------------------------------------------

async def _test():
    import sys
    company = sys.argv[1] if len(sys.argv) > 1 else "삼성전자"
    print(f"=== Naver Finance: {company} ===\n")
    r = await fetch_naver_finance(company)
    if r.error:
        print(f"Error: {r.error}")
    else:
        print(r.to_markdown())


if __name__ == "__main__":
    asyncio.run(_test())

File: src/data_sources/naver_finance.js
#!/usr/bin/env node
/**
 * naver_finance.js — Naver Finance scraper for Korean/Asian company data.
 *
 * Usage:
 *   node naver_finance.js "Company Name or Korean Ticker"
 *
 * Outputs: JSON object to stdout with financial metrics.
 * Exit codes:
 *   0 = success
 *   1 = usage error
 *   2 = scrape failed / company not found
 *
 * Called by naver_finance.py as a subprocess.
 * Useful for Korean KOSPI/KOSDAQ companies not covered by US data sources.
 */

const puppeteer = require('puppeteer');

const query = process.argv[2];
if (!query) {
  process.stderr.write('Usage: node naver_finance.js "Company Name"\n');
  process.exit(1);
}

const SEARCH_URL = `https://finance.naver.com/search/searchList.naver?query=${encodeURIComponent(query)}`;
const WAIT_MS = 3000;

async function scrape() {
  const browser = await puppeteer.launch({
    headless: true,
    args: [
      '--no-sandbox',
      '--disable-setuid-sandbox',
      '--disable-dev-shm-usage',
      '--disable-gpu',
    ],
  });

  try {
    const page = await browser.newPage();
    await page.setUserAgent(
      'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 ' +
      '(KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36'
    );

    // Step 1: Search for the company
    await page.goto(SEARCH_URL, { waitUntil: 'domcontentloaded', timeout: 20000 });
    await new Promise(r => setTimeout(r, WAIT_MS));

    // Extract first matching company's stock code and name from search results
    const firstResult = await page.evaluate(() => {
      // Naver Finance search result table: rows have company name links
      const rows = document.querySelectorAll('table.tbl_search tbody tr, .search_result table tbody tr');
      for (const row of rows) {
        const link = row.querySelector('a[href*="code="]');
        if (!link) continue;
        const href = link.href || '';
        const codeMatch = href.match(/code=(\d{6})/);
        if (!codeMatch) continue;
        return {
          code: codeMatch[1],
          name: link.textContent.trim(),
        };
      }
      return null;
    });

    if (!firstResult) {
      process.stderr.write(`No results found for "${query}" on Naver Finance.\n`);
      process.exit(2);
    }

    // Step 2: Navigate to the company's main page
    const companyUrl = `https://finance.naver.com/item/main.naver?code=${firstResult.code}`;
    await page.goto(companyUrl, { waitUntil: 'domcontentloaded', timeout: 20000 });
    await new Promise(r => setTimeout(r, WAIT_MS));

    // Step 3: Extract financial data
    const data = await page.evaluate((code, name) => {
      const getText = (sel) => {
        const el = document.querySelector(sel);
        return el ? el.textContent.trim() : null;
      };

      // Current price — typically in .no_today .blind or .stock_price
      const priceEl = document.querySelector('.no_today .blind') ||
                      document.querySelector('#_nowVal') ||
                      document.querySelector('.stock_price');
      const price = priceEl ? priceEl.textContent.replace(/[^0-9,]/g, '').replace(/,/g, '') : null;

      // Key stats table — Naver Finance shows PER, PBR, ROE in a table
      const stats = {};
      const statRows = document.querySelectorAll('.tab_con1 table tr, .section_trade table tr, table.tb_type1 tr');
      for (const row of statRows) {
        const cells = row.querySelectorAll('th, td');
        if (cells.length >= 2) {
          const key = cells[0].textContent.trim();
          const val = cells[1].textContent.trim();
          if (key && val) stats[key] = val;
        }
      }

      // Market cap — often labeled 시가총액
      const marketCapEl = document.querySelector('[class*="cap"], [class*="market"]');

      // Extract from aside/summary tables
      const tableData = {};
      const allTables = document.querySelectorAll('table');
      for (const tbl of allTables) {
        const rows = tbl.querySelectorAll('tr');
        for (const row of rows) {
          const ths = row.querySelectorAll('th');
          const tds = row.querySelectorAll('td');
          for (let i = 0; i < Math.min(ths.length, tds.length); i++) {
            const k = ths[i].textContent.trim();
            const v = tds[i].textContent.trim();
            if (k && v && k.length < 20) tableData[k] = v;
          }
        }
      }

      return {
        stock_code: code,
        company_name: name,
        current_price_krw: price ? parseInt(price) : null,
        stats: stats,
        table_data: tableData,
        page_url: window.location.href,
      };
    }, firstResult.code, firstResult.name);

    // Clean and structure the output
    const result = {
      stock_code: data.stock_code,
      company_name: data.company_name,
      exchange: 'KOSPI/KOSDAQ',
      current_price_krw: data.current_price_krw,
      page_url: data.page_url,
      metrics: {},
    };

    // Map Korean labels to English keys
    const koreanMap = {
      'PER(배)': 'per',
      'PBR(배)': 'pbr',
      'ROE(%)': 'roe_pct',
      '시가총액': 'market_cap_display',
      '매출액': 'revenue_display',
      '영업이익': 'operating_income_display',
      '당기순이익': 'net_income_display',
      '부채비율': 'debt_ratio_pct',
      '배당수익률': 'dividend_yield_pct',
      'EPS(원)': 'eps_krw',
    };

    const combined = { ...data.stats, ...data.table_data };
    for (const [kor, eng] of Object.entries(koreanMap)) {
      if (combined[kor]) {
        result.metrics[eng] = combined[kor];
      }
    }

    process.stdout.write(JSON.stringify(result, null, 2));
    process.exit(0);

  } finally {
    await browser.close();
  }
}

scrape().catch(err => {
  process.stderr.write(`Naver Finance scrape error: ${err.message}\n`);
  process.exit(2);
});
